### ISSUE TITLE
Object (/v1, Kind=Endpoints: test-endpoint) keeps changing, diffing without lock

### ISSUE THREAD

**Issue by T4cC0re** [kubernetes/kubernetes#107980]
### What happened?

When diffing an `Endpoints` object with `--server-side=true`, there is no diff produced, but this message is printed:

```
W0207 05:09:34.901634   96076 diff.go:692] Object (/v1, Kind=Endpoints: test-endpoint) keeps changing, diffing without lock
Error from server (Conflict): Apply failed with 1 conflict: conflict with "kubectl-client-side-apply" using v1: .subsets
```

A diff without `--server-side=true` yields empty output as expected

### What did you expect to happen?

I expected there to be simply no diff. A client-side diff reports no changes, as expected.

### How can we reproduce it (as minimally and precisely as possible)?

test.yml: 
```yaml
---
apiVersion: v1
kind: Endpoints
metadata:
  labels:
    tanka.dev/environment: 85d1b5ffe5bfdd8a4aae0983528822b49fd733dbe02b7d09
  name: test-endpoint
  namespace: apps
subsets:
- addresses:
  - ip: 192.168.178.2
  ports:
  - port: 9999
```

```
kubectl apply -f test.yml
kubectl diff -f test.yml --server-side=true
```

### Anything else we need to know?

If the `test.yml` gets **applied** via `--server-side=true`, there is no error, but the `time` field changes, so this is undesirable in my case.

The yaml is generated via tanka 0.20, but I double-checked the output against Kubernetes documentation, and it seems perfectly valid.

If I remember correctly this was working back when I was using `minikube` version `1.20.0` (with whichever Kubernetes version was the default of that). I was also using tanka with that, and the same `Endpoins` config.

### Kubernetes version

<details>
```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.3", GitCommit:"816c97ab8cff8a1c72eccca1026f7820e93e0d25", GitTreeState:"clean", BuildDate:"2022-01-25T21:17:57Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"darwin/arm64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.6+k3s1", GitCommit:"3228d9cb9a4727d48f60de4f1ab472f7c50df904", GitTreeState:"clean", BuildDate:"2022-01-25T01:27:44Z", GoVersion:"go1.16.10", Compiler:"gc", Platform:"linux/amd64"}
```
</details>

### Cloud provider

<details>
No cloud provider. Simple one-node k3s install without config changes.
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.3 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.3 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

$ uname -a
Linux rarity.home.t4cc0.re 5.4.0-91-generic #102-Ubuntu SMP Fri Nov 5 16:31:28 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by apelisse** [comment]
Yeah, there are a few things that I don't understand here. 

What is the problem you're concerned about:
1. the poor interaction between client-side and server-side?
2. The warning about "keeps changing, diffing without lock"?
3. The time change?

**Comment by T4cC0re** [comment]
> What is the reason to use diff with server-side=true, even if you already applied using client-side?

@ardaguclu This is what tanka does internally to compare the local state (rendered as yaml) with the k8s API. This works on every other object, except for `Endpoints`-Objects. And this worked in a previous version of Kubernetes, so I believe this is a regression.

> Yeah, there are a few things that I don't understand here.
> 
> What is the problem you're concerned about:
> 
> 1. the poor interaction between client-side and server-side?
> 2. The warning about "keeps changing, diffing without lock"?
> 3. The time change?

@apelisse I would expect a `kubectl diff -f test.yml --server-side=true` to properly diff the change, instead of erroring out. It does so for every other Object. My comments about an apply with `server-side=true` were just to illustrate the fact, that in that case, the diffing of the `subsets` field works fine if the resource was created with that flag.

This has worked before and stopped working somewhere between minikube 1.20 (I believe this was kubernetes 1.21 or 1.20) and Kubernetes 1.22.6.

I don't know how the internals work, but I would guess, that something goes wrong when diffing the `subsets` array, which causes it to not be a stable comparison.

**Comment by apelisse** [comment]
I don't know how this would have worked before 1.20, did we have server-side apply at that time? For sure we know that client-side and server-side integration aren't great.

**Comment by jcleira** [comment]
👋  Not sure if this would be helpful, but I had a pre 1.20 minikube and kubectl version and I just tested this:

```bash
➜  bin git:(206bcadf021) ✗ ./kubectl version
Client Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.1-dirty", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"dirty", BuildDate:"2022-02-14T16:29:35Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"206bcadf021e76c27513500ca24182692aabd17e", GitTreeState:"clean", BuildDate:"2020-09-09T11:18:22Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
➜  bin git:(206bcadf021) ✗ ./kubectl apply -f ~/test.yml
endpoints/test-endpoint configured
➜  bin git:(206bcadf021) ✗ ./kubectl diff -f ~/test.yml --server-side=true

W0214 17:47:02.679297   66169 diff.go:528] Object (/v1, Kind=Endpoints: test-endpoint) keeps changing, diffing without lock
Error from server (Conflict): Apply failed with 1 conflict: conflict with "kubectl-client-side-apply" using v1: .subsets
➜  bin git:(206bcadf021) ✗
```

**Comment by T4cC0re** [comment]
Thanks for testing this @jcleira . I might be misremembering then, so it is definitely not a regression. Sorry for that confusion.

Regardless, I would still assume the diff to not fail at this point :)

**Comment by jcleira** [comment]
I got to have a look more in detail on the diff command under `--server-side=true`.

Using the flag is triggering an `apply` attempt for the endpoint BEFORE doing the `diff`, and the conflict is the error returned because the initial `apply` is a client-side apply and the second one (triggered by adding the flag as mentioned) is `server-side`.

It would work with this command chain:
```bash
kubectl apply -f test.yaml --server-side=true
kubectl diff -f test.yml --server-side=true
```

Probably we could inverse the order and do the `diff` then the `apply`, so the error message would be something like:

```bash
diff -u -N /var/folders/y_/sv0954yd5cb51y9l2w8v8c840000gn/T/LIVE-2264261010/v1.Endpoints.default.test-endpoint /var/folders/y_/sv0954yd5cb51y9l2w8v8c840000gn/T/MERGED-1608721413/v1.Endpoints.default.test-endpoint
--- /var/folders/y_/sv0954yd5cb51y9l2w8v8c840000gn/T/LIVE-2264261010/v1.Endpoints.default.test-endpoint	2022-03-01 17:34:52.000000000 +0100
+++ /var/folders/y_/sv0954yd5cb51y9l2w8v8c840000gn/T/MERGED-1608721413/v1.Endpoints.default.test-endpoint	2022-03-01 17:34:52.000000000 +0100
@@ -14,7 +14,7 @@
       f:subsets: {}
     manager: kubectl
     operation: Apply
-    time: "2022-03-01T16:34:40Z"
+    time: "2022-03-01T16:34:52Z"
   name: test-endpoint
   namespace: default
   resourceVersion: "31703"
W0301 17:21:36.876011   92434 diff.go:692] Object (/v1, Kind=Endpoints: test-endpoint) keeps changing, diffing without lock
Error from server (Conflict): Apply failed with 1 conflict: conflict with "kubectl-client-side-apply" using v1: .subsets
```

WDYT? I could do a PR for that
